nalgebra-sparse/src/cs.rs

```diff
@@ -558,14 +558,21 @@ where
     T: Scalar,
 {
     let mut values_option = values;
-    if values_option.is_none() {
-        if sort {
+
+    if sort {
+        if let Some(values) = values_option.as_mut() {
+            if minor_indices.len() != values.len() {
+                return Err(SparseFormatError::from_kind_and_msg(
+                    SparseFormatErrorKind::InvalidStructure,
+                    "Number of values and minor indices must be the same.",
+                ));
+            }
```

Comment:
I think, last commit did not really restrict the conditions for comparing the lengths of `values` and `minor_indices` but even extended them a bit. Previously, the check was only done if `sort == true` and `monotonic == false` (see previous [parent if-block](https://github.com/dimforge/nalgebra/blob/d7e7dcd9b5c09f194eff9ac6ae717fcf4400d457/nalgebra-sparse/src/cs.rs#L650)).
The problem is not clear to me. This check is only required if sorting is expected in current method. Otherwise it does not matter whether the lengths are same or not as no sparse matrix creation happens in this method.
Maybe something could break in another method, where `values` are needed and no corresponding check is performed but probably not here. 
Of course if someone modifies current method and uses `values` without extending the check conditions, it might lead to undesired behavior. But then a corresponding unit-test should be added as for every change anywhere else.

However, I do not have an overview on the whole package and the planned changes.
Do you think this check should be done independently from the value of `sort`?

Reply:
You're right, I didn't realize that the previous check was inside of several blocks (part of the motivation for making it explicit up front, by the way!).

Well, my understanding is that at *present*, the problematic code path that I outlined cannot be entered. However, I think the logic is a bit confusing. For example, it's **crucial** that if `values` are not `None`, then we must **never** invalidate the invariant that `values.len() == minor_indices.len()` otherwise we risk UB in other parts of the code that may rely on this invariant. This is really independent if `sort == true` or not. So I'd really prefer to have this switched around, e.g.
```rust
if let Some(values) = values_option.as_mut() {
    if minor_indices.len() != values.len() {
        return Err(SparseFormatError::from_kind_and_msg(
            SparseFormatErrorKind::InvalidStructure,
            "Number of values and minor indices must be the same.",
        ));
    }
} else if sort {
    unreachable!(
        "Internal error: Sorting currently not supported if no values are present."
    );
}
```
Not sure if I'm bringing my message across very well here: basically the first time I read the check I was confused because we start with checking `sort` - but this invariant in principle has nothing to do with sorting. I agree that the end result is at present pretty much the same, however.

Sorry for being so pedantic here, but I think it's very important to make the invariant checks here crystal clear for anyone interpreting the code later.